### ISSUE TITLE
[SPI] TX fifo wrong status returned

### ISSUE THREAD

**Issue by e-puerto** [lowRISC/opentitan#17644]
### Description

after resetting the Spi_host IP 
- if I write into TXDATA register  a u8 (or a u32) then I read the STATUS register, the TXEMPTY field is always set and the TXQD is 0.
- if I write twice into TXDATA register  then the TXEMPTY is 0 and TXQD is 1

test made on FPGA cw310.

**Comment by adurbin-rivos** [comment]
@msfschaffner

**Comment by msfschaffner** [comment]
@johngt would you mind looping in the appropriate folks for SPI_HOST?

**Comment by marnovandermaas** [comment]
Thanks for raising this. I'll have a look at what the documentation and RTL say about this behavior. It would be useful to have instructions on how to replicate this behavior in case you have them.

**Comment by marnovandermaas** [comment]
Ok, I ran a test to see what was going on. I was running on commit: 4ce7b81ba with the following command:
```
util/dvsim/dvsim.py hw/ip/spi_host/dv/spi_host_sim_cfg.hjson -i spi_host_event -r 1 --waves shm --tool xcelium
```
From this I got the following wave trace, which confirms the behavior you mention:
![image](https://user-images.githubusercontent.com/34654485/235245533-596d00fb-4c23-4c80-b2b2-e3b6515d41aa.png)

Digging a bit further, this is because the `TXDATA` FIFO is set up to be pass through when it is empty:
https://github.com/lowRISC/opentitan/blob/319d32e45b1d72bb98990bdff83653a90410e9a5/hw/ip/spi_host/rtl/spi_host_data_fifos.sv#L77
This means that the first write you do actually gets passed through that FIFO and instead gets latched inside the `spi_host_core`, inside `spi_host_byte_select` and in a `prim_packer_fifo`.

So from a hardware point of view this is fine. We can delay the popping of the `TXFIFO` by disabling pass-through, but that would only delay the problem for 1 cycle. Another approach is to leave the data that is currently being sent by the core inside the `TXFIFO` and only pop it once the core is ready for the next one.

Personally, I think we should leave the behavior as it is unless there is an urgent software need for the meaning of `TXEMPTY` and `TXQD` to change.

**Comment by a-will** [comment]
To be clear, it's not because of first-word-fall-through in `prim_fifo_sync`, right? The depth should be accurate there. It looks like for the "pass through" case, it still increments the wptr and uses storage in the array. So the Pass parameter is irrelevant, thankfully. Otherwise, this would be a major, cross-IP concern! :smile: 

The cause is that there is _untracked_ storage in `prim_packer_fifo` that greedily pops data from the FIFO in question. This is confusing, at least, but perhaps it would be enough to document the behavior. It's slightly unnerving that there could be partial state left behind that software can't view, but at the moment, I can't think of a specific use case with a known bad outcome.

**Comment by johngt** [comment]
@moidx / @msfschaffner / @GregAC - if we are ok with the behaviour as detailed above would suggest that this gets deferred into the backlog for now rather than get an RTL change and change back to a P1 once we have gone through the other RTL priority items.

**Comment by msfschaffner** [comment]
Agreed - let's document this behavior for now.
We can fix this after M2.5.

**Comment by marnovandermaas** [comment]
I created a PR to document this behavior: https://github.com/lowRISC/opentitan/pull/18317

**Comment by rivos-eblot** [comment]
> We can fix this after M2.5.

As this ticket is now closed, what would be the best way to track this?

**Comment by msfschaffner** [comment]
We can just reopen the bug and put it in the Milestone backlog.

**Comment by msfschaffner** [comment]
Are we ok with keeping this errata, or is there straightforward fix for this?

**Comment by rivos-eblot** [comment]
I think this means that the SW has no way to know if there's something or not that may be sent on the next SPI TX sequence (unless it actively tracks what has been done and it is able to recover from some error). Is there a way to reset the SPI host locally, _i.e._ without relying on an external component - from a SW perspective - such as the reset manager?

I guess that this means the SW should not use TX_EMPTY nor TXQD in any case.

**Comment by rivos-eblot** [comment]
I was mostly thinking about error/trouble recovery/tests: it would be nice if possible to know whether before a new SPI TX transaction is started to know whether some unexpected left over byte is going to be sent to the remote device. Testing TX_EMPTY is one way, resetting the SPI host maybe another. Once the transfer is started, it is likely already too late.
